vizro-core/tests/tests_utils/demo_asserts.py

```diff
@@ -0,0 +1,46 @@
+"""Demo to show how to use asserts. These are not real tests that are run as part of testing, just a teaching aid."""
+from typing import List
+
+from asserts import assert_component_equal
+from dash import html
+
+from vizro.models import VizroBaseModel
+
+
+class X(VizroBaseModel):
+    # Low-level contents model.
+    text: str
+
+    def build(self):
+        return html.Div(
+            [html.H1("Heading"), html.P(self.text, id=self.id), html.Hr(), html.H2("Something")], className="inner"
+        )
+
+
+class Y(VizroBaseModel):
+    # Higher-level container model.
+    children: List[X]
+
+    def build(self):
+        return html.Div([child.build() for child in self.children], id=self.id, className="container")
+
+
+def test_X_build():
+    # Test for low-level contents: compare the whole component tree.
+    # Sometimes setting keys_to_strip={"className"} is useful here.
+    result = X(id="x", text="Hello world").build()
+    expected = html.Div(
+        [html.H1("Heading"), html.P("Hello world", id="x"), html.Hr(), html.H2("Something")], className="inner"
+    )
+    assert_component_equal(result, expected)
+
+
+def test_Y_build():
+    # Test for higher-level container.
+    many_x = [X(text="Hello world") for _ in range(4)]
+    result = Y(id="y", children=many_x).build()
+    # We don't want to compare the whole component tree here. Instead compare the bit that's specifically in Y and
+    # ignore the children:
+    assert_component_equal(result, html.Div(id="y", className="container"), keys_to_strip={"children"})
+    # And also compare the "interface" between X and Y:
+    assert all(isinstance(child, html.Div) for child in result.children)
```

Comment:
Just for me to when we apply this everywhere: In this case where you want to test for the higher-level component and therefore ignore its children, would you always add this additional check then:

`assert all(isinstance(child, html.Div) for child in result.children)`

or should we actually not care at all as the children should be tested properly in the tests for the children?

Reply:
Or rather in the test for `X` in this case no? Are we not sure due to pydantic and the tests for `X` that this is correct?

Reply:
I'm glad you asked because I actually meant to rewrite this and then forgot... Now done in [85dae66](https://github.com/mckinsey/vizro/pull/195/commits/85dae66647274a914ebe5d3d4273ac4b6068368c) by introducing a new `STRIP_ALL` option.

In general any test where we have `keys_to_strip={"children"}` would be followed by a test on `result.children` using `STRIP_ALL`.

The idea is:
1. `test_X_build` tests stuff that's only in `X`
2. `test_Y_build` tests stuff that's only in `Y` 
3. `test_Y_build` _also_  tests that `Y.build()` is correctly calling `X.build()` but not the details of what `X.build()` returns

The typical way you'd do case 3 is by providing a mock `X` and then doing something like `assert X.call_count == 4`. The current solution is arguably slightly "wrong" in that some of the implementation of `X.build` is leaking into the `test_Y_build`, because we are checking against `html.Div`. This is fine by me though, because it means we test the interface a bit more strictly, and it's much clearer, easier and less error-prone than writing mocks.

Reply:
Got it! That makes sense - could you add it to the PR description? 🙏  Just in case this conversation gets resolved and we try to find it when applying these changes elsewhere 👍